File: src/services/siteSearch.ts
/**
 * Site-Specific Search Service
 * 
 * Provides targeted search across research-focused sites:
 * YouTube, ArXiv, GitHub, Reddit, Wikipedia, Semantic Scholar,
 * Hacker News, Stack Overflow, PubMed, and any site via Brave fallback.
 */

import { API_BASE_URL } from './api';

const API_BASE = API_BASE_URL;

export interface SearchResult {
  title: string;
  url: string;
  snippet: string;
  source_site: string;
  published_date?: string;
  author?: string;
  thumbnail?: string;
  metadata?: Record<string, any>;
}

export interface SiteSearchResponse {
  query: string;
  site_domain?: string;
  time_range: string;
  results: SearchResult[];
  total_results: number;
}

export interface SupportedSite {
  domain: string;
  name: string;
  requires_api_key: boolean;
  api_key_env_var?: string;
}

export type TimeRange = 'all' | '24h' | '7d' | '14d' | '30d' | '90d' | '1y';

export const TIME_RANGE_OPTIONS: { value: TimeRange; label: string }[] = [
  { value: 'all', label: 'All Time' },
  { value: '24h', label: 'Last 24 Hours' },
  { value: '7d', label: 'Last 7 Days' },
  { value: '14d', label: 'Last 2 Weeks' },
  { value: '30d', label: 'Last 30 Days' },
  { value: '90d', label: 'Last 90 Days' },
  { value: '1y', label: 'Last Year' },
];

export const siteSearchService = {
  /**
   * Get list of sites with native search support
   */
  async getSupportedSites(): Promise<SupportedSite[]> {
    const response = await fetch(`${API_BASE}/site-search/supported-sites`);
    if (!response.ok) throw new Error('Failed to get supported sites');
    return response.json();
  },

  /**
   * Search a specific site or the web
   * 
   * @param query - Search query
   * @param siteDomain - Optional site domain (e.g., "youtube.com")
   * @param timeRange - Time filter
   * @param maxResults - Maximum results to return
   */
  async search(
    query: string,
    siteDomain?: string,
    timeRange: TimeRange = 'all',
    maxResults: number = 10
  ): Promise<SiteSearchResponse> {
    const response = await fetch(`${API_BASE}/site-search/search`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        query,
        site_domain: siteDomain,
        time_range: timeRange,
        max_results: maxResults,
      }),
    });
    
    if (!response.ok) throw new Error('Search failed');
    return response.json();
  },

  /**
   * Parse a search input to extract site: prefix
   * 
   * Examples:
   * - "youtube.com quantum computing" -> { site: "youtube.com", query: "quantum computing" }
   * - "quantum computing" -> { site: undefined, query: "quantum computing" }
   */
  parseSearchInput(input: string): { site?: string; query: string } {
    // Check for domain-like prefix followed by space or tab
    const domainMatch = input.match(/^([a-z0-9][-a-z0-9]*(?:\.[a-z0-9][-a-z0-9]*)+)[\s\t]+(.+)$/i);
    
    if (domainMatch) {
      return {
        site: domainMatch[1].toLowerCase(),
        query: domainMatch[2].trim(),
      };
    }
    
    return { query: input.trim() };
  },

  /**
   * Check if a domain is in our supported sites list
   */
  isSupportedSite(domain: string, supportedSites: SupportedSite[]): boolean {
    const normalized = domain.toLowerCase();
    return supportedSites.some(s => 
      s.domain === normalized || 
      s.domain.includes(normalized) ||
      normalized.includes(s.domain)
    );
  },
};

File: backend/api/site_search.py
"""Site-Specific Search API endpoints

Provides targeted search across research-focused sites.
Supports: YouTube, ArXiv, GitHub, Reddit, Wikipedia, Semantic Scholar,
          Hacker News, Stack Overflow, PubMed, and any site via Brave fallback.
"""

from typing import List, Optional
from fastapi import APIRouter, Query
from pydantic import BaseModel, Field

from services.site_search import (
    site_search_service, 
    TimeRange
)


router = APIRouter(prefix="/site-search", tags=["site-search"])


# =============================================================================
# Request/Response Models
# =============================================================================

class SiteSearchRequest(BaseModel):
    query: str = Field(description="Search query")
    site_domain: Optional[str] = Field(
        default=None, 
        description="Site domain to search (e.g., 'youtube.com'). If omitted, searches all web."
    )
    time_range: str = Field(
        default="all",
        description="Time filter: all, 24h, 7d, 14d, 30d, 90d, 1y"
    )
    max_results: int = Field(default=10, ge=1, le=50)


class SearchResultResponse(BaseModel):
    title: str
    url: str
    snippet: str
    source_site: str
    published_date: Optional[str] = None
    author: Optional[str] = None
    thumbnail: Optional[str] = None
    metadata: Optional[dict] = None


class SiteSearchResponse(BaseModel):
    query: str
    site_domain: Optional[str]
    time_range: str
    results: List[SearchResultResponse]
    total_results: int


class SupportedSite(BaseModel):
    domain: str
    name: str
    requires_api_key: bool
    api_key_env_var: Optional[str] = None


# =============================================================================
# API Endpoints
# =============================================================================

@router.get("/supported-sites", response_model=List[SupportedSite])
async def get_supported_sites():
    """Get list of sites with native search support."""
    return site_search_service.get_supported_sites()


@router.post("/search", response_model=SiteSearchResponse)
async def search(request: SiteSearchRequest):
    """
    Search a specific site or the web.
    
    Examples:
    - Site search: {"query": "quantum computing", "site_domain": "youtube.com", "time_range": "30d"}
    - Web search: {"query": "quantum computing", "time_range": "7d"}
    """
    # Parse time range
    time_range_map = {
        "all": TimeRange.ALL_TIME,
        "24h": TimeRange.LAST_24H,
        "7d": TimeRange.LAST_7D,
        "14d": TimeRange.LAST_14D,
        "30d": TimeRange.LAST_30D,
        "90d": TimeRange.LAST_90D,
        "1y": TimeRange.LAST_YEAR,
    }
    
    time_range = time_range_map.get(request.time_range, TimeRange.ALL_TIME)
    
    # Execute search
    results = await site_search_service.search(
        query=request.query,
        site_domain=request.site_domain,
        time_range=time_range,
        max_results=request.max_results
    )
    
    # Convert to response format
    result_responses = [
        SearchResultResponse(
            title=r.title,
            url=r.url,
            snippet=r.snippet,
            source_site=r.source_site,
            published_date=r.published_date,
            author=r.author,
            thumbnail=r.thumbnail,
            metadata=r.metadata
        )
        for r in results
    ]
    
    return SiteSearchResponse(
        query=request.query,
        site_domain=request.site_domain,
        time_range=request.time_range,
        results=result_responses,
        total_results=len(result_responses)
    )


@router.get("/search")
async def search_get(
    q: str = Query(..., description="Search query"),
    site: Optional[str] = Query(None, description="Site domain to search"),
    time: str = Query("all", description="Time filter: all, 24h, 7d, 14d, 30d, 90d, 1y"),
    limit: int = Query(10, ge=1, le=50, description="Max results")
):
    """
    GET endpoint for site search (convenience for browser testing).
    
    Examples:
    - /site-search/search?q=quantum+computing&site=youtube.com&time=30d
    - /site-search/search?q=machine+learning&site=arxiv.org
    """
    request = SiteSearchRequest(
        query=q,
        site_domain=site,
        time_range=time,
        max_results=limit
    )
    return await search(request)
